acra-censor/handlers/whitelist_handler.go

```diff
@@ -65,6 +81,107 @@ func (handler *WhitelistHandler) CheckQuery(query string) (bool, error) {
 	return false, nil
 }
 
+func (handler *WhitelistHandler) handleAliasedTables(parsedQuery sqlparser.TableExprs) error {
+	allowedTablesCounter := 0
+	var err error
+	for _, allowedTable := range handler.tables {
+		for _, table := range parsedQuery {
+			switch table.(type) {
+			case *sqlparser.AliasedTableExpr:
+				if strings.EqualFold(sqlparser.String(table.(*sqlparser.AliasedTableExpr).Expr), allowedTable) {
+					allowedTablesCounter++
+				}
+				break
+			case *sqlparser.JoinTableExpr:
+				err = handler.handleJoinedTables(table.(*sqlparser.JoinTableExpr))
+				if err != nil {
+					return ErrAccessToForbiddenTableWhitelist
+				}
+				break
+			case *sqlparser.ParenTableExpr:
+				err = handler.handleParenTables(table.(*sqlparser.ParenTableExpr))
+				if err != nil {
+					return ErrAccessToForbiddenTableWhitelist
+				}
+				break
+			default:
+				return ErrUnexpectedTypeError
+			}
+
+			if err != nil {
+				return ErrAccessToForbiddenTableWhitelist
+			}
+		}
+	}
+
+	if allowedTablesCounter != len(parsedQuery) {
+		return ErrAccessToForbiddenTableWhitelist
```

Comment:
can you add unit test for this method where `allowedTablesCounter != len(parsedQuery)`
i'm not sure that it works correctly because you use local `allowedTablesCounter` variable that will not incremented with other types of expressions than `AliasedTableExpr`. And in case when query will contain not only aliased expressions you always will have error

Reply:
Can you please provide example of sql query for this unit-test? I thought that existing unit tests are enough (`testWhitelistTables` and `TestDifferentTablesParsing`)

Reply:
It seems that `allowedTablesCounter` is global for inner functions that are called recursively. This behaviour is similar: https://play.golang.org/p/KoVkUb5ydrt

Reply:
yes, it's global for inner functions. but here we don't create inner function with closure. this situation similar to https://play.golang.org/p/j7-hUWDxcu9

Reply:
plus better to rewrite with map usage instead arrays for table validation and exclude extra outer for-loop